sqlx-macros-core/src/derives/decode.rs

```diff
@@ -265,24 +265,21 @@ fn expand_derive_decode_struct(
     if cfg!(feature = "postgres") {
         let ident = &input.ident;
 
-        // extract type generics
-        let generics = &input.generics;
-        let (_, ty_generics, _) = generics.split_for_impl();
+        let (_, ty_generics, where_clause) = input.generics.split_for_impl();
 
-        // add db type for impl generics & where clause
-        let mut generics = generics.clone();
-        generics.params.insert(0, parse_quote!('r));
-
-        let predicates = &mut generics.make_where_clause().predicates;
-
-        for field in fields {
-            let ty = &field.ty;
+        let mut generics = input.generics.clone();
 
-            predicates.push(parse_quote!(#ty: ::sqlx::decode::Decode<'r, ::sqlx::Postgres>));
-            predicates.push(parse_quote!(#ty: ::sqlx::types::Type<::sqlx::Postgres>));
+        // add db type for impl generics & where clause
+        for type_param in &mut generics.type_params_mut() {
+            type_param.bounds.extend::<[TypeParamBound; 2]>([
+                parse_quote!(for<'decode> ::sqlx::decode::Decode<'decode, ::sqlx::Postgres>),
```

Comment:
Why is this now a higher-order bound instead of just `Decode<'r, _>` like before?

Reply:
honestly, i wrote this so long ago, i don't recall any of the reasoning behind my changes. happy to change this though

Reply:
If it works, I think `'r` without the `for<>` makes more sense. If not, I'd be curious why.

Reply:
It may have been to get around an "implementation is not general enough" error. People complain about those from time to time.

This change is what I was thinking about in this comment BTW https://github.com/launchbadge/sqlx/pull/2940?notification_referrer_id=NT_kwDOADDOg7I4Nzk3Mjc1MTk0OjMxOTg1OTU#issuecomment-1977999552

Reply:
Which "this comment"? GitHub doesn't highlight anything for me when clicking that link.

And I'm pretty sure the `for<'decode>` makes the impl *less* general and can't possibly fix any errors.

Reply:
The comment I @-mention'd you in above.

Reply:
Ahh, so you agree with keeping the `'r` if possible, right?